lib/open_food_network/scope_variants_for_search.rb

```diff
@@ -83,28 +85,33 @@ def scope_variants_to_distributor(variants, distributor)
     end
 
     def variant_in_stock_sql
-      [
-        variant_override_in_stock_exists_sql,
-        spree_stock_item_in_stock_exists_sql
-      ].reject(&:blank?).join(" OR ")
+      "EXISTS (#{variant_override_in_stock_sql}) OR (
+        NOT EXISTS(#{variant_override_not_in_stock_sql}) AND
+        EXISTS(#{spree_stock_item_in_stock_sql})
+      )"
     end
 
-    def spree_stock_item_in_stock_exists_sql
-      return if distributor.present?
+    def spree_stock_item_in_stock_sql
+      "SELECT 1
+       FROM spree_stock_items
+       WHERE variant_id = spree_variants.id AND (backorderable = true OR count_on_hand > 0)"
+    end
 
-      "EXISTS(
-         SELECT 1
-         FROM spree_stock_items
-         WHERE variant_id = spree_variants.id AND (backorderable = true OR count_on_hand > 0)
-      )"
+    def variant_override_in_stock_sql
+      "SELECT 1
+       FROM variant_overrides
+       WHERE variant_id = spree_variants.id AND
+             hub_id = #{distributor.id} AND
+             (on_demand = true OR count_on_hand > 0)"
     end
 
-    def variant_override_in_stock_exists_sql
-      sql = "SELECT 1
-             FROM variant_overrides
-             WHERE variant_id = spree_variants.id AND (on_demand = true OR count_on_hand > 0)"
-      sql += " AND hub_id = #{distributor.id}" if distributor.present?
-      "EXISTS (#{sql})"
+    def variant_override_not_in_stock_sql
+      "SELECT 1
+       FROM variant_overrides
+       WHERE variant_id = spree_variants.id AND
+             hub_id = #{distributor.id} AND
+             on_demand = false AND
+             count_on_hand = 0"
     end
```

Comment:
We are getting closer! But I find the code very difficult to understand. I'm trying to simplify it here:

You main equation is:

- VO in stock OR
- no VO out of stock AND item in stock

The "VO in stock" part is correct. The "no VO out of stock" seems to be a clever way of solving this but I'm still wondering if it's the same as "no VO for stock present". The original logic of variant overrides was: use the variant's value unless the variant override has a non-null value to override it. So let's go through it case by case:

* The variant override is missing. Your `variant_override_not_in_stock_sql` would be empty and we look at the variant's stock. :heavy_check_mark: 
* The variant override is present, on_demand is `false` and the count_on_hand is `0`. Your `variant_override_not_in_stock_sql` would find it and we ignore that variant. :heavy_check_mark: 
* The variant override is present, on_demand and count_on_hand are both `NULL`. Your `variant_override_not_in_stock_sql` would be empty and we look at the variant's stock. :heavy_check_mark: 
* The variant override is present, on_demand is `true`. We don't find that and look at the variant's stock. :x: 
* The variant override is present, count_on_hand is positive.  We don't find that and look at the variant's stock. :x: 

So we need specs for the last two cases.

And instead of "no VO out of stock" we we can check if the variant override is either missing or has null values.

Alternatively, you could join the tables and query it like this:

```
vo.on_demand = true OR vo.count_on_hand > 0 OR
 (vo.on_demand is null AND stock_item.backorderable = true) OR
 (vo.count_on_hand is null AND stock_item.count_on_hand > 0)
```

Does that sound logical?

There's a nice table of values in the validation code:
https://github.com/openfoodfoundation/openfoodnetwork/blob/43f28d94b9bc4905ac55d1de4869a0d89260d679/app/models/concerns/stock_settings_override_validation.rb#L11-L18